lib/V8/processMonitor.cpp

```diff
@@ -0,0 +1,170 @@
+#include <sys/wait.h>
+#endif
+
+#ifdef TRI_HAVE_MACH
+#include <mach/mach_host.h>
+#include <mach/mach_port.h>
+#include <mach/mach_traps.h>
+#include <mach/task.h>
+#include <mach/thread_act.h>
+#include <mach/vm_map.h>
+#endif
+
+#ifdef _WIN32
+#include <Psapi.h>
+#include <TlHelp32.h>
+#include "Basics/socket-utils.h"
+#include "Basics/win-utils.h"
+#endif
+#include <fcntl.h>
+
+#ifdef TRI_HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#include "Basics/Mutex.h"
+#include "Basics/MutexLocker.h"
+#include "Basics/DeadlockDetector.h"
+#include "Basics/Thread.h"
+#include "V8/processMonitor.h"
+#include "V8/v8-deadline.h"
+
+using namespace arangodb;
+
+namespace {
+
+/// @brief lock for protected access to vector ExternalProcesses
+static arangodb::Mutex ExitedExternalProcessesLock;
+
+}  // namespace
+
+namespace arangodb {
+
+/// @brief if monitored & exited we keep them here:
+std::map<TRI_pid_t, ExternalProcessStatus> ExitedExternalProcessStatus;
+
+std::vector<ExternalId> monitoredProcesses;
+
+void addMonitorPID(ExternalId& pid) {
+  MUTEX_LOCKER(mutexLocker, ExitedExternalProcessesLock);
+  monitoredProcesses.push_back(pid);
+}
+
+void removeMonitorPID(ExternalId const& pid) {
+  MUTEX_LOCKER(mutexLocker, ExitedExternalProcessesLock);
+  for (auto it = monitoredProcesses.begin(); it != monitoredProcesses.end();
+       ++it) {
+    if (it->_pid == pid._pid) {
+      monitoredProcesses.erase(it);
+      break;
+    }
+  }
+}
+
+std::optional<ExternalProcessStatus> getHistoricStatus(TRI_pid_t pid) {
+  MUTEX_LOCKER(mutexLocker, ExitedExternalProcessesLock);
+  auto it = ExitedExternalProcessStatus.find(pid);
+  if (it == ExitedExternalProcessStatus.end()) {
+    return std::nullopt;
+  }
+  return std::optional<ExternalProcessStatus>{it->second};
+}
+
+class ProcessMonitorThread : public arangodb::Thread {
+ public:
+  ProcessMonitorThread(application_features::ApplicationServer& server)
+      : Thread(server, "ProcessMonitorThread") {}
+  ~ProcessMonitorThread() { shutdown(); }
+
+ protected:
+  void run() override {
+    while (!isStopping()) {
+      std::vector<ExternalId> mp;
+      {
+        MUTEX_LOCKER(mutexLocker, ExitedExternalProcessesLock);
+        mp = monitoredProcesses;
+      }
+      for (auto const& pid : mp) {
+        auto status = TRI_CheckExternalProcess(pid, false, 0);
+        if ((status._status == TRI_EXT_TERMINATED) ||
+            (status._status == TRI_EXT_ABORTED) ||
+            (status._status == TRI_EXT_NOT_FOUND)) {
+          // Its dead and gone - good
+          removeMonitorPID(pid);
+          ExitedExternalProcessStatus[pid._pid] = status;
+          triggerV8DeadlineNow(false);
+        }
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+  }
+};
+
+namespace {
+ProcessMonitorThread* mt = nullptr;
+}
+
+void launchMonitorThread(application_features::ApplicationServer& server) {
+  if (mt == nullptr) {
+    mt = new ProcessMonitorThread(server);
```

Comment:
Please use `std::make_unique`.
Even better: instead of having these global functions and another global variable `mt`, why not move the entire management of the ProcessMonitor thread into its own ApplicationFeature?